src/core/instance/events.js

```diff
@@ -38,13 +38,18 @@ export function updateComponentListeners (
 
 export function eventsMixin (Vue: Class<Component>) {
   const hookRE = /^hook:/
-  Vue.prototype.$on = function (event: string, fn: Function): Component {
+  Vue.prototype.$on = function (event: string | Array<string>, fn: Function): Component {
     const vm: Component = this
-    ;(vm._events[event] || (vm._events[event] = [])).push(fn)
-    // optimize hook:event cost by using a boolean flag marked at registration
-    // instead of a hash lookup
-    if (hookRE.test(event)) {
-      vm._hasHookEvent = true
+    const events: Array<string> = Array.isArray(event) ? event : [event]
```

Comment:
I'd prefer using for loop only when it's an Array instead of always turning it into an Array.

In the loop you can just do `this.$on(events[i], fn)`